actors/builtin/miner/partition_state_test.go

```diff
@@ -0,0 +1,591 @@
+		ExpectBQ().
+			Add(terminationEpoch, 3, 5).
+			Equals(t, queue)
+
+		// pop the rest
+		result, hasMore, err = partition.PopEarlyTerminations(adt.AsStore(rt), 5)
+		require.NoError(t, err)
+
+		// expect 3 and 5
+		assertBitfieldsEqual(t, bf(3, 5), result.Sectors[terminationEpoch])
+
+		// expect no more results
+		assert.False(t, hasMore)
+
+		// expect early terminations to be empty
+		queue, err = miner.LoadBitfieldQueue(adt.AsStore(rt), partition.EarlyTerminated, quantSpec)
+		require.NoError(t, err)
+		ExpectBQ().Equals(t, queue)
+	})
+}
+
+type expectExpirationGroup struct {
+	expiration abi.ChainEpoch
+	sectors    *bitfield.BitField
+}
+
+func assertPartitionExpirationQueue(t *testing.T, rt *mock.Runtime, partition *miner.Partition, quant miner.QuantSpec, groups []expectExpirationGroup) {
+	queue, err := miner.LoadExpirationQueue(adt.AsStore(rt), partition.ExpirationsEpochs, quant)
+	require.NoError(t, err)
+
+	for _, group := range groups {
+		requireNoExpirationGroupsBefore(t, group.expiration, queue)
+		set, err := queue.PopUntil(group.expiration)
+		require.NoError(t, err)
+
+		// we pnly care whether the sectors are in the queue or not. ExpirationQueue tests can deal with early or on time.
+		allSectors, err := bitfield.MergeBitFields(set.OnTimeSectors, set.EarlySectors)
+		require.NoError(t, err)
+		assertBitfieldsEqual(t, group.sectors, allSectors)
+	}
+}
+
+func assertPartitionState(t *testing.T,
+	partition *miner.Partition,
+	sectors []*miner.SectorOnChainInfo,
+	sectorSize abi.SectorSize,
+	allSectorIds *bitfield.BitField,
+	faults *bitfield.BitField,
+	recovering *bitfield.BitField,
+	terminations *bitfield.BitField) {
+
+	faultyPower := miner.PowerForSectors(sectorSize, selectSectors(t, sectors, faults))
+	assert.True(t, partition.FaultyPower.Equals(faultyPower))
+	recoveringPower := miner.PowerForSectors(sectorSize, selectSectors(t, sectors, recovering))
+	assert.True(t, partition.RecoveringPower.Equals(recoveringPower))
+	livePower := miner.PowerForSectors(sectorSize, sectors)
+	assert.True(t, partition.LivePower.Equals(livePower))
+	activePower := livePower.Sub(faultyPower)
+	assert.True(t, activePower.Equals(partition.ActivePower()))
+
+	assertBitfieldsEqual(t, faults, partition.Faults)
+	assertBitfieldsEqual(t, recovering, partition.Recoveries)
+	assertBitfieldsEqual(t, terminations, partition.Terminated)
+	assertBitfieldsEqual(t, allSectorIds, partition.Sectors)
+
+	live, err := partition.LiveSectors()
+	require.NoError(t, err)
+
+	expectedLive, err := bitfield.SubtractBitField(allSectorIds, terminations)
+	require.NoError(t, err)
+	assertBitfieldsEqual(t, expectedLive, live)
+
+	nonActiveIds, err := bitfield.MergeBitFields(faults, terminations)
+	require.NoError(t, err)
+	expectedActiveIds, err := bitfield.SubtractBitField(allSectorIds, nonActiveIds)
+	require.NoError(t, err)
+
+	active, err := partition.ActiveSectors()
+	require.NoError(t, err)
+	assertBitfieldsEqual(t, expectedActiveIds, active)
+}
```

Comment:
There are some state invariants that should be maintained. Even though probably covered by these assertions, could we check them explicitly? Perhaps even add a "checkInvariants" method to the state object for use by tests.
- faults, recoveries, terminated are contained in sectors
- recoveries are contained in faults
- terminated is exclusive with faults
- all sectors are in expiration queue exactly once
- early terminated is subset of terminated sectors

Reply:
I created a `checkPartitionInvariants` function instead of extending the partition state. We can further split these functions up but this probably covers everything for now.